### ISSUE TITLE
Suggestion on JavaScript support

### ISSUE THREAD

**Issue by aloneguid** [sjdirect/abot#91]
First of all I'm crawling pages with Abot for a few months and it's great, thanks for this project and please keep the work going, it's awesome!

I've encountered a problem with a few massive and important (for my business) websites that render links with JavaScript which Abot ignores with built-in features for very obvious reasons. To solve that I've created an extension which uses Awesomium project (Chrome runtime) and that works very well as it does execute JS locally and speed wise it's not too far from HttpWebRequest, obviously there is an overhead, but it's bearable.

I wouldn't like to keep it in my private repo and want to contribute to Abot. I was thinking of creating a new project alongside your solution which people can reference to replace the built-in crawler - Awesomium has lots of external dependencies, no nuget package and is a bit heavy. Can you please suggest the best way to add this into Abot?

**Comment by mikegron** [comment]
fyi we implemented a similar solution but using the headless browser PhantomJS, since it does not open a browser window and is pretty lightweight (http://phantomjs.org/), and manipulating it using the C# selenium webdriver for it (https://github.com/SeleniumHQ/selenium).
Basically, we are overridding the WebContentExtrator and PageRequester to use this web driver, so the changes are minimal, we only pass the custom classes to the WebCrawler ctor.

**Comment by obender** [comment]
Can any of you please share the code, and I will integrate the best solution in proper manner into the project, 

Generally  if there a missing Nuget package of the dependency you just can create it, and add a parameter to the configuration to change the crawler provider.

**Comment by LeMoussel** [comment]
I think that instead of using C# selenium webdriver, it must be possible to use [PhantomJS Inter Process Communication](http://phantomjs.org/inter-process-communication.html). I think that the overhead would be less than with selenium webdriver.

**Comment by sjdirect** [comment]
Hi,
The most ideal situation would be to have each implementation as a separate plugin with a nuget package. I don't think having Abot source code in with these is a good idea. 

Ie.. A typical user that wants Abot and the ability to render js would... 

1) Nuget install "Abot" using nuget,
2) Nuget install "AbotPlugin-JSRender-Awesomium" or "AbotPlugin-JSRender-PhantomJS"
3) Manually plugin the renderer impl into PoliteWebCrawler instance

After we get to that point, I could add a cleaner hook in Abot to plug it in easier and possibly add config values if necessary.

Step 1 would be just to get a working example on a public repo somewhere so we can coordinate a bit. If you would rather not spend the time creating a repo/nuget/etc.. then you can just email me a working example in zip form.

Thanks for contributing!

**Comment by aloneguid** [comment]
@mikegron I'd really like to see your code if you don't mind sharing, I've been considering PhantomJS which I think runs on Mozilla runtime (I can be wrong) but discarded for two reasons - I found it extremely hard to configure one-click deployment on Azure Cloud Service instance, and I personally use Chrome for daily word so it was more familiar.

Awesomium comes with it's own caveats - it requires process running in 32bit (I've heard 64bit support is coming) and it can crash as bad as native libraries crash bringing down the whole node. For that reason I had to use external 32bit process with IPC (WCF TCP) to control the crawler. If it crashes I don't care as I can relaunch the process, and in addition to that I can run the server in 64bit. Moreover Awesomium redistributes as .msm module out of the box and installs dependencies in GAC and some native libraries in Program Files folder. I haven't got to the point of trying to package it without installing anything on target node, so it runs on a Windows Server VM instead of a self contained package which can be deployed to the grid (Azure Cloud Service or Service Fabric).

I didn't get to the point of making it all nice, and I agree with @sjdirect that it should be a separate Nuget package which can be manually plugged in. That would keep the core libraries clean. We could share some code on the IPC, crash recovery controlling browser instances etc., it may be same or very similar.

I'll start working on this though. Thanks everyone.

**Comment by mikegron** [comment]
I created a simple example project using some of our code. It uses the Nuget packages for Abot, PhantomJS and the selenium WebDriver.
This example has some drawbacks though, it only uses a single PhantomJS process so it is not thread-safe. Normally, you would want a PhantomJS process for every worker thread which is not hard to implement. Also, the requests are somewhat doubled since we are executing the HttpWebRequest in the PageRequester and then retrieving the content using PhantomJS, which executes another request on the server. This is because there is no way to intercept the responses from Phantom to reject pages in error (for instance, 404 not found), unless you use a proxy, but that's pretty complicated.

Zip file:
https://drive.google.com/file/d/0B1Cc-jmpA6imejgzb3J2OS1LQkE/view?usp=sharing

**Comment by sjdirect** [comment]
ok guys, i've put both of your examples here...

https://github.com/sjdirect/abot-phantomjs
https://github.com/sjdirect/abot-awesomium

@mikegron I put the phantomjs under an mit license. Please let me know if you are ok with that. @aloneguid I'm sure you can see that I forked your repo and made some mods which i have submitted a pull request to you. 

I thought this would be the best way to allow you both (and anyone else whose interested) to be able to quickly clone and tinker around with both of these solutions. The phantomjs example is a console app and the awesomium is a unit test assembly.

**Comment by aloneguid** [comment]
Thanks @sjdirect  and @mikegron. Sorry I didn't have a chance to make a huge progress, I'm working on awesomium plugin now in order to support more than one crawl instance and stress test that as much as I can.

**Comment by aloneguid** [comment]
Hey guys, I almost gave up due to the NuGet dependency hell. NuGet doesn't play well with a mixture of native/managed dlls, or when it does, upstream projects don't get the binaries to the target build folder. Whatever I do I can't make CI transparent enough and embed Awesomium, which makes it almost useless.

On the other hand, we decided that we can't use Abot in the crawling solution these days, unfortunately. We used Abot for some time and it was great, but due to overcomplicated nature of our crawlers and today's web most of the cool Abot features don't play well when it comes to speed and vast differences of rendering engines.

I'll keep trying to work on this, however I may give up and build a framework specifically for Awesomium which simply accepts site base URL on input and produces a zip archive on output.
